### PR TITLE
[Merged by Bors] - api, events, proposal: stream created and included proposals

### PR DESCRIPTION
**Pull request by dshulyak**
proposals api is used in system tests to assert common functionality (proposals are equally created, hare creates proposals, beacon is the same on all proposals)

depends on https://github.com/spacemeshos/api/pull/162

## Changes
- fire events when proposals are created or included in the block
- add proposals stream to debug service, potentially it can be moved into "public" service
- remove old unused code from events module, i don't think that we have any plan on using it and it is a bit messy

### CODE REVIEW

File: api/grpcserver/debug_service.go [hunk 1]
```diff
@@ -65,3 +69,61 @@ func (d DebugService) Accounts(_ context.Context, in *empty.Empty) (*pb.Accounts
...
+	var (
+		eventch <-chan interface{}
+		fullch  <-chan struct{}
+	)
+	if sub := events.SubcribeProposals(); sub != nil {
+		eventch, fullch = consumeEvents(stream.Context(), sub)
+	}
+	if eventch == nil || fullch == nil {
+		return status.Errorf(codes.FailedPrecondition, "event reporting is not enabled")
+	}
```
Comment (nkryuchkov) [inline api/grpcserver/debug_service.go]: `consumeEvents` always returns non-nil channels, so `codes.FailedPrecondition` would mean that `events.SubcribeProposals() == nil`, maybe we can simplify this?
```suggestion
	sub := events.SubcribeProposals()
	if sub == nil {
		return status.Errorf(codes.FailedPrecondition, "event reporting is not enabled")
	}
	eventch, fullch := consumeEvents(stream.Context(), sub)
```

**Review by nkryuchkov — APPROVED** [review]

File: events/proposals.go [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+}
+
+// SubcribeProposals subscribes to the proposals.
+func SubcribeProposals() Subscription {
+	mu.RLock()
+	defer mu.RUnlock()
+	if reporter != nil {
+		sub, err := reporter.bus.Subscribe(new(EventProposal))
+		if err != nil {
+			log.With().Panic("Failed to subscribe to proposals")
```
Comment (countvonzero) [inline events/proposals.go:54]: how is this called? i am wondering whether panic is the right thing to do here.

File: api/grpcserver/debug_service.go [hunk 3]
```diff
@@ -65,3 +69,61 @@ func (d DebugService) Accounts(_ context.Context, in *empty.Empty) (*pb.Accounts
 func (d DebugService) NetworkInfo(ctx context.Context, _ *empty.Empty) (*pb.NetworkInfoResponse, error) {
 	return &pb.NetworkInfoResponse{Id: d.identity.ID().String()}, nil
 }
+
+// ProposalsStream streams all proposals confirmed by hare.
+func (d DebugService) ProposalsStream(_ *emptypb.Empty, stream pb.DebugService_ProposalsStreamServer) error {
```
Comment (countvonzero) [inline api/grpcserver/debug_service.go:74]: where can i find definition and pb files for DebugService_ProposalsStreamServer?

Comment (dshulyak) [reply api/grpcserver/debug_service.go:74]: https://github.com/spacemeshos/api/pull/162

File: events/proposals.go (hunk 2 above)
Comment (dshulyak) [reply events/proposals.go:54]: yes this is a right thing todo, all the errors condition are based on the input from the programmer https://github.com/libp2p/go-eventbus/blob/master/basic.go#L164-L225

File: api/grpcserver/debug_service.go (hunk 1 above)
Comment (dshulyak) [reply api/grpcserver/debug_service.go]: thanks, applied

**Review by countvonzero — APPROVED** [review]

**Comment by dshulyak** [comment]
bors merge

**Pull request closed without merging**